### PR TITLE
OvS DPDK VA

### PR DESCRIPTION
**Pull request by Jaganathancse**
This change is to add VA OvS DPDK implementation.

### CODE REVIEW

File: examples/va/nfv/ovs-dpdk/edpm/README.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+## Steps
+
+1. Create Secrets
+```bash
+oc apply -f dataplanesshsecret.yaml -f nova_migration_ssh_key.yaml
+```
+2. Create OvS DPDK ConfigMaps and associated OpenStackDataPlaneService
+```bash
+oc apply -f nova_sriov.yaml
```
Comment (christophefontaine) [inline examples/va/nfv/ovs-dpdk/edpm/README.md:20]: oc apply -f nova_ovs_dpdk.yaml

File: examples/va/nfv/ovs-dpdk/edpm/openstackdataplanenodeset.yaml [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+        # CHANGEME
+        timesync_ntp_servers:
+          - hostname: clock.redhat.com
+        # CPU pinning settings
+         # edpm nfv ovs dpdk config
+         edpm_kernel_args: "default_hugepagesz=1GB hugepagesz=1G hugepages=64 iommu=pt intel_iommu=on tsx=off isolcpus=2-11,14-23"  # CHANGEME
+         edpm_tuned_profile: "cpu-partitioning"  # CHANGEME
+         edpm_tuned_isolated_cores: "2-11,14-23"  # CHANGEME
+         edpm_ovs_dpdk_pmd_core_list: "1,13,2,14,3,15"  # CHANGEME
+         edpm_ovs_dpdk_lcore_list: "0,12"  # CHANGEME
```
Comment (christophefontaine) [inline examples/va/nfv/ovs-dpdk/edpm/openstackdataplanenodeset.yaml:122]: lcore list should be removed.

Comment (Jaganathancse) [reply examples/va/nfv/ovs-dpdk/edpm/openstackdataplanenodeset.yaml:122]: > lcore list should be removed.

Ack

File: examples/va/nfv/ovs-dpdk/edpm/README.md (hunk 1 above)
Comment (Jaganathancse) [reply examples/va/nfv/ovs-dpdk/edpm/README.md:20]: > oc apply -f nova_ovs_dpdk.yaml

Ack

File: examples/va/nfv/ovs-dpdk/edpm/openstackdataplanenodeset.yaml (hunk 2 above)
Comment (Jaganathancse) [reply examples/va/nfv/ovs-dpdk/edpm/openstackdataplanenodeset.yaml]: @christophefontaine done

Comment (Jaganathancse) [reply examples/va/nfv/ovs-dpdk/edpm/openstackdataplanenodeset.yaml]: @christophefontaine
NeutronPhysnetNUMANodesMapping` and `NeutronTunnelNUMANodes` parameters configuration are missing now,
NeutronPhysnetNUMANodesMapping parameter config,
[neutron]
physnets = dpdk1, dpdk2

[neutron_physnet_dpdk1]
numa_nodes = 0

[neutron_physnet_dpdk2
numa_nodes = 0,1

NeutronTunnelNUMANodes config,
[neutron_tunnel]
numa_nodes = 1

do we need to add these parameters also like cpu pinnig?

File: examples/va/nfv/ovs-dpdk/README.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
+# OpenStack OvS DPDK
+
+**Based on OpenStack K8S operators from the "main" branch of the [OpenStack Operator repo](https://github.com/openstack-k8s-operators/openstack-operator/tree/12c57baeca4ae33dd30a7707d330eb094309b4cd) on Dec 6th, 2023**
```
Comment (abays) [inline examples/va/nfv/ovs-dpdk/README.md:3]: Is this true or just a copy-paste from a previous VA?

File: examples/va/nfv/ovs-dpdk/edpm/nova_ovs_dpdk.yaml [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: ovs-dpdk-cpu-pinning-nova
+data:
+  04-cpu-pinning-nova.conf: |
+    [compute]
+    cpu_shared_set = 0-3,24-27 # CHANGEME
+    cpu_dedicated_set = 8-23,32-47 # CHANGEME
+    reserved_host_memory_mb = 4096
```
Comment (abays) [inline examples/va/nfv/ovs-dpdk/edpm/nova_ovs_dpdk.yaml:18]: This format is not correct as per information received from the compute team.  Please see this for reference: https://github.com/openstack-k8s-operators/architecture/blob/b18ae4a41c1bf27d58b15a17d8b7b95db4931ed2/examples/va/nfv/sriov/edpm/values.yaml#L137-L146

File: examples/va/nfv/ovs-dpdk/edpm/openstackdataplanenodeset.yaml [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+        service_net_map:
+          nova_api_network: internal_api
+          nova_libvirt_network: internal_api
+        # CHANGEME
+        timesync_ntp_servers:
+          - hostname: clock.redhat.com
+        # CPU pinning settings
+         # edpm nfv ovs dpdk config
+         edpm_kernel_args: "default_hugepagesz=1GB hugepagesz=1G hugepages=64 iommu=pt intel_iommu=on tsx=off isolcpus=2-11,14-23"  # CHANGEME
+         edpm_tuned_profile: "cpu-partitioning"  # CHANGEME
```
Comment (abays) [inline examples/va/nfv/ovs-dpdk/edpm/openstackdataplanenodeset.yaml:117]: We are using this [1] for SRIOV as per the recommendation of the compute team.  Should we use the same here?

[1] https://github.com/openstack-k8s-operators/architecture/blob/b18ae4a41c1bf27d58b15a17d8b7b95db4931ed2/examples/va/nfv/sriov/edpm/values.yaml#L32

File: examples/va/nfv/ovs-dpdk/kustomization.yaml [hunk 6]
```diff
@@ -0,0 +1,12 @@
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+
+components:
+  - ../../../../va/nfv/sriov/
```
Comment (abays) [inline examples/va/nfv/ovs-dpdk/kustomization.yaml:5]: This is pointing to the SRIOV VA and needs to point to an underlying DPDK VA.  See [1] for SRIOV and then perhaps use that to inspire something similar for DPDK.

[1] https://github.com/openstack-k8s-operators/architecture/tree/main/va/nfv/sriov

Comment (Jaganathancse) [reply examples/va/nfv/ovs-dpdk/kustomization.yaml:5]: Ack. Thanks will update it.

File: examples/va/nfv/ovs-dpdk/edpm/nova_ovs_dpdk.yaml (hunk 4 above)
Comment (Jaganathancse) [reply examples/va/nfv/ovs-dpdk/edpm/nova_ovs_dpdk.yaml:18]: @abays Thanks.
reserved_huge_pages config needed?
Can you please corresponding parameters in tripleo?

Comment (Jaganathancse) [reply examples/va/nfv/ovs-dpdk/edpm/nova_ovs_dpdk.yaml:18]: > This format is not correct as per information received from the compute team. Please see this for reference:
> 
> https://github.com/openstack-k8s-operators/architecture/blob/b18ae4a41c1bf27d58b15a17d8b7b95db4931ed2/examples/va/nfv/sriov/edpm/values.yaml#L137-L146


Ack. reserved_host_memory_mb is part of DEFAULT section only. will update it.

Comment (abays) [reply examples/va/nfv/ovs-dpdk/edpm/nova_ovs_dpdk.yaml:18]: Unfortunately I don't know what the related OOO parameters are.  @SeanMooney helped me understand what the proper approach is here [1].

[1] https://github.com/openstack-k8s-operators/architecture/pull/76#discussion_r1456407995

File: examples/va/nfv/ovs-dpdk/README.md (hunk 3 above)
Comment (Jaganathancse) [reply examples/va/nfv/ovs-dpdk/README.md:3]: @abays yes just copied. updated date based on my PR's.

> Is this true or just a copy-paste from a previous VA?

@abays yes just copied. updated date based on my PR's.

File: examples/va/nfv/ovs-dpdk/edpm/openstackdataplanenodeset.yaml (hunk 5 above)
Comment (Jaganathancse) [reply examples/va/nfv/ovs-dpdk/edpm/openstackdataplanenodeset.yaml:117]: > We are using this [1] for SRIOV as per the recommendation of the compute team. Should we use the same here?
> 
> [1]
> 
> https://github.com/openstack-k8s-operators/architecture/blob/b18ae4a41c1bf27d58b15a17d8b7b95db4931ed2/examples/va/nfv/sriov/edpm/values.yaml#L32

done

File: examples/va/nfv/ovs-dpdk/kustomization.yaml (hunk 6 above)
Comment (Jaganathancse) [reply examples/va/nfv/ovs-dpdk/kustomization.yaml:5]: > Ack. Thanks will update it.

done

File: examples/va/nfv/ovs-dpdk/edpm/openstackdataplanenodeset.yaml (hunk 2 above)
Comment (Jaganathancse) [reply examples/va/nfv/ovs-dpdk/edpm/openstackdataplanenodeset.yaml]: > Yes we'll need these parameters. If we can have the same syntax everywhere, that would be great, so we don't have an intermediate parameter such as NeutronPhysnetNUMANodesMappingandNeutronTunnelNUMANodes
> 
> So, would something like the following work ?
> 
> ```
> neutron:
>      customServiceConfig: |
>           [neutron]
>           physnets = dpdk1, dpdk2
>           [neutron_physnet_dpdk1]
>           numa_nodes = 0
>           [neutron_physnet_dpdk2]
>           numa_nodes = 0,1
>           [neutron_tunnel]
>           numa_nodes = 1
> ```

done

File: examples/va/nfv/ovs-dpdk/README.md (hunk 3 above)
Comment (Jaganathancse) [reply examples/va/nfv/ovs-dpdk/README.md]: > Is this true or just a copy-paste from a previous VA?

done

File: examples/va/nfv/ovs-dpdk/README.md [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+## Considerations
+
+1. These CRs are validated for the overall functionality of the OSP cloud deployed, but they nonetheless require customization for the particular environment in which they are utilized.  In this sense they are _templates_ meant to be consumed and tweaked to fit the specific constraints of the hardware available.
+
+2. The CRs are applied against an OpenShift cluster in _stages_.  That is, there is an ordering in which each grouping of CRs is fed to the cluster.  It is _not_ a case of simply taking all CRs from all stages and applying them all at once.
+
+3. In stages 1 and 2 [kustomize](https://kustomize.io/) is used to genereate the CRs dynamically. The `*-values.yaml` file(s) must be updated to fit your environment. kustomize version 5 or newer required.
+
+4. In stage 3 YAML comments are placed throughout the CRs to aid in the process of customizing the CRs.  Fields that _must_ (or most likely need to be) changed are commented with "# CHANGEME" either on the field itself or somewhere nearby.  Other comments are added to explain fields that can be changed and, sometimes, to explain additions that can be made.
```
[the remark this answers is not in the document]
Comment (Jaganathancse) [reply examples/va/nfv/ovs-dpdk/README.md:22]: > As a follow-up, we can attempt to add `kustomize` to the data plane portion

done

File: examples/va/nfv/ovs-dpdk/kustomization.yaml (hunk 6 above)
Comment (abays) [reply examples/va/nfv/ovs-dpdk/kustomization.yaml]: You've changed the path, but the path does not actually exist.  You will need to create it and put the appropriate kustomization logic in it.  For example, see what we did for SRIOV [1].

[1] https://github.com/openstack-k8s-operators/architecture/tree/main/va/nfv/sriov

Comment (Jaganathancse) [reply examples/va/nfv/ovs-dpdk/kustomization.yaml]: @abays Ack. will create it. Thanks.

Comment (Jaganathancse) [reply examples/va/nfv/ovs-dpdk/kustomization.yaml]: @abays done.

File: examples/va/nfv/ovs-dpdk/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (abays) [reply examples/va/nfv/ovs-dpdk/README.md:22]: I'm confused as to why you said "done" here.  It doesn't look like the `edpm` dir has any sort of kustomization.  Regardless, again, this is something we can do in a follow-up.

**Review by abays — APPROVED** [review]
/lgtm

File: examples/va/nfv/ovs-dpdk/README.md [hunk 8]
```diff
@@ -0,0 +1,30 @@
+# OpenStack OvS DPDK
+
+**Based on OpenStack K8S operators from the "main" branch of the [OpenStack Operator repo](https://github.com/openstack-k8s-operators/openstack-operator/tree/12c57baeca4ae33dd30a7707d330eb094309b4cd) on Jan 31st, 2024**
+
+This is a collection of CR templates that represent a validated Red Hat OpenStack Services on OpenShift deployment that has the following characteristics:
+
+- 3 master/worker combo-node OpenShift cluster
```
Comment (fultonj) [inline examples/va/nfv/ovs-dpdk/README.md:7]: We could put OvS DPDK on this list but I guess we're assuming that's obvious given the title of the document.

File: examples/va/nfv/ovs-dpdk/control-plane.md [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+```
+Alternatively use your own copies of those files and edit
+[kustomization.yaml](kustomization.yaml) to use those copies.
+```
+resources:
+  - values-ci-framework.yaml
+  - service-values-ci-framework.yaml
+```
+
+Generate the control-plane and networking CRs.
```
Comment (fultonj) [inline examples/va/nfv/ovs-dpdk/control-plane.md:32]: This is not correct as NNCP has been split out into a separate operation. 

https://github.com/openstack-k8s-operators/architecture/commit/e313fb6320607c6d673b3c5878e2c02ab44a8bcf

Please rebase the branch you are submitting this on against main as it is 18 commits behind. 

https://github.com/Jaganathancse/architecture/tree/ovs_dpdk_va_changes

Once you rebase apply the missing steps about nncp to your new control-plane section.

File: examples/va/nfv/ovs-dpdk/control-plane.md [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+kustomize build > control-plane.yaml
+```
+
+## Create CRs
+```
+oc apply -f control-plane.yaml
+```
+If the `oc apply` command  produces a `no matches for kind
+NodeNetworkConfigurationPolicy` error, then see
+[Missing NNCPs Workaround](../../../../docs/wa/missing_nncp.md).
```
Comment (fultonj) [inline examples/va/nfv/ovs-dpdk/control-plane.md:43]: This is also out of date. We no longer require this workaround. Please rebase and use latest updates.

File: examples/va/nfv/ovs-dpdk/edpm/openstackdataplanenodeset.yaml [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+          {% set min_viable_mtu = mtu_list | max %}
+          network_config:
+          - type: ovs_bridge
+            name: {{ neutron_physical_bridge_name }}
+            mtu: {{ min_viable_mtu }}
+            use_dhcp: false
+            dns_servers: {{ ctlplane_dns_nameservers }}
+            domain: {{ dns_search_domains }}
+            addresses:
+            - ip_netmask: {{ ctlplane_ip }}/{{ ctlplane_subnet_cidr }}
```
Comment (fultonj) [inline examples/va/nfv/ovs-dpdk/edpm/openstackdataplanenodeset.yaml:145]: This is also out  of date. Please update so you are in-line with.

https://github.com/Jaganathancse/architecture/commit/5a8b7d665b981b1483452f906cd2464f04dec0b8

File: examples/va/nfv/ovs-dpdk/values.yaml [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+      - allocationRanges:
+          - end: 172.17.0.250
+            start: 172.17.0.100
+        cidr: 172.17.0.0/24
+        name: subnet1
+        vlan: 21
+    mtu: 1500
+    prefix-length: 24
+    iface: enp7s0.20
+    vlan: 202
```
Comment (fultonj) [inline examples/va/nfv/ovs-dpdk/values.yaml:75]: Why is vlan 202 here and vlan 21 on line 71? Please make them consistent. In a future follow up when this is moved to kustomize they can both be supplied from a values.yaml file but for now let's make them consistent.

File: va/nfv/ovs-dpdk/kustomization.yaml [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+  setRoleBindingSubjects: none
+  unsetOnly: true
+  fieldSpecs:
+  - path: metadata/name
+    kind: Namespace
+    create: true
+
+components:
+- ../../../lib/deploy
+- ../../../lib/nncp
```
Comment (fultonj) [inline va/nfv/ovs-dpdk/kustomization.yaml:21]: These two lines are out of date. Please rebase and apply the changes.

https://github.com/Jaganathancse/architecture/commit/e313fb6320607c6d673b3c5878e2c02ab44a8bcf

**Review by fultonj — CHANGES_REQUESTED** [review]
Please rebase your source and apply the changes that you copied/pasted from outdated sources to the new content in this PR.

Otherwise it looks good.

File: va/nfv/ovs-dpdk/kustomization.yaml (hunk 13 above)
Comment (fultonj) [reply va/nfv/ovs-dpdk/kustomization.yaml:21]: This components list is out of date. It should only have the following:
```
components:
- ../../../lib/networking
- ../../../lib/control-plane
```

For example please see this:

https://github.com/openstack-k8s-operators/architecture/blob/main/va/nfv/sriov/kustomization.yaml#L19-L21

**Review by fultonj — COMMENTED** [review]
Looks better @Jaganathancse thought we still need one more change to the components list in va/nfv/ovs-dpdk/kustomization.yaml as commented below.

**Comment by Jaganathancse** [comment]
> Looks better @Jaganathancse though we still need one more change to the components list in va/nfv/ovs-dpdk/kustomization.yaml as commented below.

Thanks @fultonj

**Review by fultonj — APPROVED** [review]

**Pull request merged**